app/src/org/commcare/utils/EncryptionUtils.java

```
@@ -1,17 +1,274 @@
+
+        return transformation;
+    }
+
+    //Encryption backed by Android KeyStore
+    private static byte[] encrypt(Context context, byte[] bytes)
+            throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
+            IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException,
+            UnrecoverableEntryException, CertificateException, KeyStoreException, IOException, NoSuchProviderException {
+        String transformation = getTransformationString();
+        Cipher cipher = Cipher.getInstance(transformation);
+        cipher.init(Cipher.ENCRYPT_MODE, getKey(context, true));
+        byte[] encrypted = cipher.doFinal(bytes);
+        byte[] iv = cipher.getIV();
+        int ivLength = iv == null ? 0 : iv.length;
+
+        byte[] output = new byte[encrypted.length + ivLength + 3];
+        int writeIndex = 0;
+        output[writeIndex] = (byte) ivLength;
+        writeIndex++;
+        if(ivLength > 0) {
+            System.arraycopy(iv, 0, output, writeIndex, iv.length);
+            writeIndex += iv.length;
+        }
+
+        output[writeIndex] = (byte) (encrypted.length / 256);
+        writeIndex++;
+        output[writeIndex] = (byte) (encrypted.length % 256);
+        writeIndex++;
+        System.arraycopy(encrypted, 0, output, writeIndex, encrypted.length);
+
+        return output;
+    }
+
+    //Decryption backed by Android KeyStore
+    private static byte[] decrypt(Context context, byte[] bytes)
+            throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidAlgorithmParameterException,
+            InvalidKeyException, IllegalBlockSizeException, BadPaddingException,
+            UnrecoverableEntryException, CertificateException, KeyStoreException, IOException, NoSuchProviderException {
+        int readIndex = 0;
+        int ivLength = bytes[readIndex];
+        readIndex++;
+        if(ivLength < 0) {
+            //TODO: Early chance to catch decryption error
```

probably should throw a specific exception from here for callers to catch and act on

---

Done